stress: failed test in cockroach/gossip/gossip.test: TestGossipOrphanedStallDetection

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/7a59042b04a56c69dc7956b97f5dbce6b8312f13

Stress build found a failed test:

```
=== RUN   TestGossipOrphanedStallDetection
W160830 05:17:02.852395 gossip/gossip.go:1037  no resolvers found; use --join to specify a connected node
W160830 05:17:02.853916 gossip/gossip.go:1037  no resolvers found; use --join to specify a connected node
I160830 05:17:02.856401 gossip/client.go:111  node 2: started gossip client to 127.0.0.1:34454
I160830 05:17:02.857389 stopper.go:408  quiesceing; tasks left:
2      gossip/infostore.go:290
I160830 05:17:02.857458 stopper.go:408  quiesceing; tasks left:
1      gossip/infostore.go:290
I160830 05:17:02.857541 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:34454->127.0.0.1:39848: use of closed network connection
I160830 05:17:02.857562 http2_client.go:1029  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160830 05:17:02.857811 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:02.861849 gossip/gossip.go:1053  node has connected to cluster via gossip
I160830 05:17:03.857936 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:05.290911 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:07.982285 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:11.466891 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:17.447406 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:27.557251 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:43.897937 /go/src/google.golang.org/grpc/clientconn.go:667  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:34454: getsockopt: connection refused"; Reconnecting to {"127.0.0.1:34454" <nil>}
I160830 05:17:47.944167 /go/src/google.golang.org/grpc/clientconn.go:767  grpc: addrConn.transportMonitor exits due to: context canceled
--- FAIL: TestGossipOrphanedStallDetection (45.09s)
	testing.go:117: gossip/gossip_test.go:308, condition failed to evaluate within 45s: 2 still connected
```

Run Details:

```
32 runs so far, 0 failures, over 5s
66 runs so far, 0 failures, over 10s
109 runs so far, 0 failures, over 15s
149 runs so far, 0 failures, over 20s
187 runs so far, 0 failures, over 25s
225 runs so far, 0 failures, over 30s
265 runs so far, 0 failures, over 35s
299 runs so far, 0 failures, over 40s
333 runs so far, 0 failures, over 45s
367 runs so far, 0 failures, over 50s
395 runs so far, 0 failures, over 55s
426 runs so far, 0 failures, over 1m0s
452 runs completed, 1 failures, over 1m4s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client.go
+++ b/gossip/client.go
@@ -236,9 +236,9 @@ func (c *client) handleResponse(g *Gossip, reply *Response) error {
 	// matches an incoming or the client is connecting to itself.
 	if g.mu.is.NodeID == c.peerID {
 		return errors.Errorf("stopping outgoing client to node %d (%s); loopback connection", c.peerID, c.addr)
-	} else if g.hasIncomingLocked(c.peerID) && g.mu.is.NodeID < c.peerID {
+	} else if g.hasIncomingLocked(c.peerID) && g.mu.is.NodeID > c.peerID {
 		// To avoid mutual shutdown, we only shutdown our client if our
-		// node ID is less than the peer's.
+		// node ID is higher than the peer's.
 		return errors.Errorf("stopping outgoing client to node %d (%s); already have incoming", c.peerID, c.addr)
 	}
 

--- a/gossip/client.go
+++ b/gossip/client.go
@@ -19,6 +19,7 @@ package gossip
 import (
 	"fmt"
 	"net"
+	"sync"
 	"time"
 
 	"github.com/pkg/errors"
@@ -80,8 +81,19 @@ func (c *client) start(
 ) {
 	stopper.RunWorker(func() {
 		ctx, cancel := context.WithCancel(context.TODO())
-		defer cancel()
+		var wg sync.WaitGroup
 		defer func() {
+			// This closes the outgoing stream, causing any attempt to send or
+			// receive to return an error.
+			//
+			// Note: it is still possible for incoming gossip to be processed after
+			// this point.
+			cancel()
+
+			// The stream is closed, but there may still be some incoming gossip
+			// being processed. Wait until that is complete to avoid racing the
+			// client's removal against the discovery of its remote's node ID.
+			wg.Wait()
 			disconnected <- c
 		}()
 
@@ -109,7 +121,7 @@ func (c *client) start(
 
 		// Start gossiping.
 		log.Infof(ctx, "node %d: started gossip client to %s", nodeID, c.addr)
-		if err := c.gossip(ctx, g, stream, stopper); err != nil {
+		if err := c.gossip(ctx, g, stream, stopper, &wg); err != nil {
 			if !grpcutil.IsClosedConnection(err) {
 				g.mu.Lock()
 				peerID := c.peerID
@@ -248,7 +260,7 @@ func (c *client) handleResponse(g *Gossip, reply *Response) error {
 // gossip loops, sending deltas of the infostore and receiving deltas
 // in turn. If an alternate is proposed on response, the client addr
 // is modified and method returns for forwarding by caller.
-func (c *client) gossip(ctx context.Context, g *Gossip, stream Gossip_GossipClient, stopper *stop.Stopper) error {
+func (c *client) gossip(ctx context.Context, g *Gossip, stream Gossip_GossipClient, stopper *stop.Stopper, wg *sync.WaitGroup) error {
 	sendGossipChan := make(chan struct{}, 1)
 
 	// Register a callback for gossip updates.
@@ -262,7 +274,12 @@ func (c *client) gossip(ctx context.Context, g *Gossip, stream Gossip_GossipClie
 	defer g.RegisterCallback(".*", updateCallback)()
 
 	errCh := make(chan error, 1)
+	// This wait group is used to allow the caller to wait until gossip
+	// processing is terminated.
+	wg.Add(1)
 	stopper.RunWorker(func() {
+		defer wg.Done()
+
 		errCh <- func() error {
 			for {
 				reply, err := stream.Recv()

--- a/gossip/client_test.go
+++ b/gossip/client_test.go
@@ -333,7 +333,7 @@ func TestClientDisconnectRedundant(t *testing.T) {
 		// Check which of the clients is connected to the other.
 		ok1 := local.findClient(func(c *client) bool { return c.addr.String() == rAddr.String() }) != nil
 		ok2 := remote.findClient(func(c *client) bool { return c.addr.String() == lAddr.String() }) != nil
-		// We expect node 1 to disconnect; if both are still connected,
+		// We expect node 2 to disconnect; if both are still connected,
 		// it's possible that node 1 gossiped before node 2 connected, in
 		// which case we have to gossip from node 1 to trigger the
 		// disconnect redundant client code.
@@ -341,7 +341,7 @@ func TestClientDisconnectRedundant(t *testing.T) {
 			if err := local.AddInfo("local-key", nil, time.Second); err != nil {
 				t.Fatal(err)
 			}
-		} else if !ok1 && ok2 && verifyServerMaps(local, 1) && verifyServerMaps(remote, 0) {
+		} else if ok1 && !ok2 && verifyServerMaps(local, 0) && verifyServerMaps(remote, 1) {
 			return nil
 		}
 		return errors.New("local client to remote not yet closed as redundant")

